### PR TITLE
Wave D: fail-closed authorization foundation (Tier 0)

### LINKED ISSUE #37: /mcp endpoint fully unauthenticated in default config
**Severity:** Critical  ·  **Area:** Security

**Location:** `main.py:275, 318-332`; `src/core/mcp_adapter.py:37-72`

**Problem:** The MCP server is mounted at `/mcp` and, with auth off by default, all five tools (`contex_query`, `contex_publish`, `contex_create_subscription`, `contex_delete_subscription`, `read_subscription`) accept arbitrary `project_id`/`subscription_id` with no session identity and no object-level checks. The exposure is purely because no auth middleware is registered by default.

**Impact:** Any unauthenticated HTTP client can query, publish, and manage subscriptions across all projects with no identity or ownership enforcement.

**Fix:** Authenticate the MCP transport (Bearer handshake), propagate principal/tenant into handlers, and enforce ownership.

---
Source: 2026-08-22 adversarial audit (`docs/audits/2026-08-22-adversarial-audit.md`)

### PR DESCRIPTION
**Pull request by robmillersoftware**
## Description

**Wave D — Fail-Closed Authorization Foundation (Tier 0).** Makes hardened mode (`AUTH_ENABLED=true`) actually airtight. Today, even with auth on, RBAC is a total no-op on `/api/v1` (the permission table matches `/api/*` by prefix; #38) and multiple paths bypass auth entirely. This inverts the model from *allow-unless-denied* to *deny-unless-a-route-declares-otherwise*.

Design spec: `docs/superpowers/specs/2026-09-05-fail-closed-authz-foundation-design.md`
Implementation plan: `docs/superpowers/plans/2026-09-05-fail-closed-authz-foundation.md`

## Type of Change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] New feature (non-breaking change which adds functionality)

Breaking (intended, only when `AUTH_ENABLED=true`): keys with no role assignment are now **denied** rather than treated as `readonly`; every route enforces a declared permission; the deprecated `/api` alias is removed (use `/api/v1`).

## Related Issues

Closes #37, #38, #39, #40, #41, #53, #72
Relates to #42, #43, #54, #55, #73 (fast-follow ownership/IDOR checks that build on `Identity`), and #64/#68/#69/RLS (adjacent, out of scope here)

## Changes Made

- **One identity resolver** (`src/core/identity.py`): scope-authoritative `Identity` + `resolve_identity()`. Unknown/invalid key → deny (never a default role). Explicit `APIKey.scopes` win; else the role expands; roles are presets. No DB migration (reuses existing columns).
- **Route-level authz** (`src/core/authz.py`): `get_identity`, `require(*perms)`, `public` dependencies. Enforcement gated by `AUTH_ENABLED`; markers always present.
- **Boot-time coverage gate** (`src/core/authz_coverage.py`): walks `app.routes` and **refuses to boot** if any route lacks a `require()`/`public` decision (mounts + framework routes allowlisted). Runs at startup and as a CI test — this is the fail-closed keystone.
- **Every route annotated** with its required permission across all 7 routers + main.py-level routes. Permission vocabulary extended to cover webhooks / service accounts / tenants / audit / versioning.
- **MCP auth** (`src/core/mcp_adapter.py`): API-key bearer via the SDK `TokenVerifier` (no OAuth machinery), plus per-tool default-deny; tenant/project derived from token **claims**, not tool args.
- **Removed the fail-open middleware** (`rbac_middleware.py`, `APIKeyMiddleware`) and the legacy `/api` alias.
- **`X-Tenant-ID` spoofing fixed** (#41): tenant derives from the credential; a mismatched header is rejected 403.
- **Protected mode** (`src/core/protected_mode.py`): Redis-style — refuses non-loopback binds when auth is unconfigured (opt-out `CONTEX_PROTECTED_MODE=false`).

### Notes for the reviewer
- `/service-accounts/token` + `/token/validate` are intentionally `public` — they self-authenticate the service-account key in-body (gating them behind admin would chicken/egg-lock the flow).
- Rate limiting is intentionally left **unwired** — its path table had the same #38 prefix bug (already a no-op on `/api/v1`), so this is no regression; a follow-up re-adds it correctly.
- Commits are single-purpose and ordered as a narrative: spec/plan → primitives → per-router annotations → wiring/removal → MCP → protected mode → tests → review fixes.

## Testing

- [x] All existing tests pass
- [x] New tests added for new functionality
- [x] Tested edge cases

Full `tests/` suite: **422 passed, 0 failures**. New coverage: identity resolver (deny/scopes/tenant/powerless-key/legacy-scope-drop), authz deps (401/403/off), coverage gate (per-method, mounts, nested deps, real-app fully covered), MCP verifier + per-tool enforce, tenant spoofing (#41), protected mode, and an end-to-end role matrix. (~6 pre-existing `sdk/python/tests` failures are unrelated and untouched.)

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (spec/plan committed)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective
- [x] New and existing unit tests pass locally

## Additional Notes

**Post-merge follow-ups:** ownership/IDOR checks (#42/#43/#54/#55/#73) now build trivially on `Identity.tenant_id`/`has_project`; rate-limiter path fix; a project-scope check on `/sandbox/subscribe`. A few cosmetic minors are deferred (a private-symbol import, an inline import, a test-transport deprecation).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/core/authz.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    identity = await resolve_identity(request.app.state.db, credential)
+    if identity is None:
+        raise HTTPException(status_code=401, detail="Invalid API Key")
+    request.state.identity = identity
+    return identity
+
+
+def require(*permissions: Permission):
+    """Return a dependency that requires all `permissions`. Empty = authenticated-only."""
+    from fastapi import Depends
```
Comment (robmillersoftware) [inline src/core/authz.py]: Can we please do something about all of these inline imports?

File: src/core/authz.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    _dep._authz_marker = permissions  # sentinel for the coverage walker
+    return _dep
+
+
+async def public() -> None:
+    """Explicit public marker. No-op dependency; presence = intentionally unauthenticated."""
+    return None
+
+
+public._public_marker = True  # type: ignore[attr-defined]
```
Comment (robmillersoftware) [inline src/core/authz.py]: What is the point of this?

File: src/core/authz_coverage.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    while stack:
+        dep = stack.pop()
+        call = getattr(dep, "call", None)
+        # Contract (Task 3): require(*perms) always sets _authz_marker to a
+        # TUPLE, never None. An empty tuple (require() with no args =
+        # authenticated-only) is a valid, intentional decision and MUST count as
+        # covered. Hence the `is not None` presence check rather than a
+        # truthiness check — a truthiness check would silently turn the
+        # empty-tuple case into a fail-closed false-negative. Do not weaken this
+        # without also changing the marker's sentinel contract in authz.py.
```
Comment (robmillersoftware) [inline src/core/authz_coverage.py]: That's too much comment

File: src/core/authz.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        if not set(permissions).issubset(identity.scopes):
+            missing = sorted(p.value for p in set(permissions) - identity.scopes)
+            raise HTTPException(
+                status_code=403,
+                detail={"error": "forbidden", "missing_permissions": missing},
+            )
+        return identity
+
+    _dep._authz_marker = permissions  # sentinel for the coverage walker
+    return _dep
```
Comment (robmillersoftware) [inline src/core/authz.py]: I don't understand this. This is some pretty gnarly code.

File: src/core/identity.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    scopes=frozenset(Permission),
+    tenant_id=None,
+    projects=(),
+    role=Role.ADMIN,
+)
+
+
+async def resolve_identity(db: DatabaseManager, raw_key: str) -> Identity | None:
+    """Resolve a raw bearer credential to an Identity, or None (deny)."""
+    if not raw_key or not raw_key.startswith("ck_"):
```
Comment (robmillersoftware) [inline src/core/identity.py]: What is this "ck_" check?

File: src/core/identity.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+
+def _parse_scopes(raw: list[str] | None) -> frozenset[Permission]:
+    """Parse stored scope strings into Permissions, ignoring unknown/legacy values."""
+    out = set()
+    for s in raw or []:
+        try:
+            out.add(Permission(s))
+        except ValueError:
+            continue  # legacy scope strings like "read"/"write" are ignored
```
Comment (robmillersoftware) [inline src/core/identity.py]: Please don't leave these types of meta comments. It's just confusing to come across later.

File: src/core/mcp_adapter.py [hunk 7]
```diff
@@ -5,13 +5,54 @@
...
 
 
-def build_mcp_server(engine):
+def _enforce(permission, project_id=None):
+    """Default-deny per-tool check. No-op when auth is off (demo parity)."""
+    if not auth_enabled():
+        return
+    tok = get_access_token()
+    if tok is None or permission.value not in tok.scopes:
+        raise PermissionError(f"Missing permission: {permission.value}")
```
Comment (robmillersoftware) [inline src/core/mcp_adapter.py]: This doesn't reach the user, does it? We don't want to advertise our permission scheme

File: src/core/mcp_adapter.py [hunk 8]
```diff
@@ -44,12 +96,14 @@ async def contex_query(project_id: str, query: str, top_k: int = 5, threshold: f
...
     async def contex_create_subscription(project_id: str, needs: list[str],
                                          top_k: int = 5, threshold: float | None = None) -> str:
+        _enforce(Permission.QUERY_DATA, project_id=project_id)
         e = _get_engine()
         sub_id = await e.subscriptions.create(project_id, needs, top_k=top_k, threshold=threshold)
         return json.dumps({"subscription_id": sub_id, "resource_uri": f"contex://subscriptions/{sub_id}"})
 
     @server.tool(name="contex_delete_subscription", description="Delete a subscription.")
     async def contex_delete_subscription(subscription_id: str) -> str:
+        _enforce(Permission.QUERY_DATA)  # subscription ownership check is fast-follow #73
```
Comment (robmillersoftware) [inline src/core/mcp_adapter.py]: Please don't leave this sort of comment.

File: src/core/protected_mode.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
+"""Redis-style protected mode: refuse remote binds when auth is unconfigured."""
```
Comment (robmillersoftware) [inline src/core/protected_mode.py]: Could we not advertise for Redis?

File: src/core/tenant_middleware.py [hunk 10]
```diff
@@ -167,25 +181,35 @@ async def _identify_tenant(
...
+        When authentication is disabled (demo mode), the legacy header/API-key/
+        path precedence is used unchanged.
 
         Returns:
             Tenant ID if identified, None otherwise
+
+        Raises:
+            _TenantSpoofingError: if auth is on and X-Tenant-ID disagrees with
+                the identity's tenant.
         """
```
Comment (robmillersoftware) [inline src/core/tenant_middleware.py:189]: This seems like an excessively long docstring

File: src/core/tenant_middleware.py [hunk 11]
```diff
@@ -167,25 +181,35 @@ async def _identify_tenant(
...
+        from src.core.authz import auth_enabled
+
+        if auth_enabled():
+            return await self._identify_tenant_from_identity(request)
+
+        # --- Demo mode (auth disabled): legacy header-based behavior ---
```
Comment (robmillersoftware) [inline src/core/tenant_middleware.py]: There are multiple references to "legacy" behavior in this file. Why are we keeping legacy code?

File: src/core/tenant_middleware.py [hunk 12]
```diff
@@ -201,6 +225,44 @@ async def _identify_tenant(
...
+        is not yet populated here. We therefore resolve the credential directly
+        (a cheap indexed hash lookup). This does NOT return 401 for a
+        missing/invalid credential — that remains the single responsibility of
+        the route's ``get_identity`` dependency. We simply decline to set a
+        tenant and let the request proceed to be rejected downstream.
+
+        If the credential resolves to an identity and the client also sent an
+        ``X-Tenant-ID`` that does not match ``identity.tenant_id``, this raises
+        :class:`_TenantSpoofingError`.
+        """
```
Comment (robmillersoftware) [inline src/core/tenant_middleware.py]: This can be way more concise

File: src/core/tenant_middleware.py [hunk 13]
```diff
@@ -201,6 +225,44 @@ async def _identify_tenant(
...
+        from src.core.authz import _extract_credential
+        from src.core.identity import resolve_identity
```
Comment (robmillersoftware) [inline src/core/tenant_middleware.py]: Inline imports

File: src/core/authz.py (hunk 1 above)
Comment (robmillersoftware) [reply src/core/authz.py]: Hoisted `Depends` to module top-level — `require`/`public` are classes now, so nothing is imported inside the function. (54080e6)

File: src/core/authz.py (hunk 2 above)
Comment (robmillersoftware) [reply src/core/authz.py]: This marks a route as intentionally public so the boot-time coverage gate recognizes it. Refactored: `public` is now a `_Public` singleton the gate detects by identity, not a monkeypatched attribute. (54080e6)

File: src/core/authz_coverage.py (hunk 3 above)
Comment (robmillersoftware) [reply src/core/authz_coverage.py]: Removed — with the `isinstance` check the mechanism is self-explanatory, so the 7-line note is gone. (54080e6)

File: src/core/authz.py (hunk 4 above)
Comment (robmillersoftware) [reply src/core/authz.py]: Rewrote `require` as a small class that stores `permissions` as a plain attribute, instead of a closure with a monkeypatched sentinel. (54080e6)

File: src/core/identity.py (hunk 5 above)
Comment (robmillersoftware) [reply src/core/identity.py]: Contex keys are minted as `ck_<token>`, so this rejects non-Contex credentials before hashing + a DB hit. Pulled the prefix into a named `_KEY_PREFIX` constant. (2dfa1cf)

File: src/core/identity.py (hunk 6 above)
Comment (robmillersoftware) [reply src/core/identity.py]: Removed the meta comment, and dropped the "legacy" wording from the docstring too. (2dfa1cf)

File: src/core/mcp_adapter.py (hunk 7 above)
Comment (robmillersoftware) [reply src/core/mcp_adapter.py]: Confirmed — the MCP SDK returns `str(e)` to the client, so it did leak the permission name. Now raises a generic `Permission denied`; also made the REST 403 generic (`Forbidden`, no missing-permissions list). (6f84abb + 54080e6)

File: src/core/mcp_adapter.py (hunk 8 above)
Comment (robmillersoftware) [reply src/core/mcp_adapter.py]: Removed. (98d500c)

File: src/core/protected_mode.py (hunk 9 above)
Comment (robmillersoftware) [reply src/core/protected_mode.py]: Reworded the docstring to drop the Redis reference. (2dfa1cf)

File: src/core/tenant_middleware.py (hunk 10 above)
Comment (robmillersoftware) [reply src/core/tenant_middleware.py:189]: Trimmed to a couple of lines. (2dfa1cf)

File: src/core/tenant_middleware.py (hunk 11 above)
Comment (robmillersoftware) [reply src/core/tenant_middleware.py]: Cleaned up: removed the dead `api_key_id` branch (nothing sets it since APIKeyMiddleware was removed) and dropped the "legacy" framing. The auth-off path is just demo-mode tenant selection (X-Tenant-ID header or `/t/` prefix). (2dfa1cf)

File: src/core/tenant_middleware.py (hunk 12 above)
Comment (robmillersoftware) [reply src/core/tenant_middleware.py]: Condensed the docstring. (2dfa1cf)

File: src/core/tenant_middleware.py (hunk 13 above)
Comment (robmillersoftware) [reply src/core/tenant_middleware.py]: Hoisted to top-level (verified no circular import); also made `extract_credential` public since it is now used cross-module. (2dfa1cf)
